streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamTaskTest.java

```diff
@@ -431,12 +407,108 @@ public void flush() {
+        ((ProcessorContextImpl) task.processorContext()).setCurrentNode(processor);
+        task.punctuate(processor, 10);
+    }
+
+    @Test
+    public void shouldCallPunctuateOnPassedInProcessorNode() throws Exception {
+        task.punctuate(processor, 5);
+        assertThat(processor.punctuatedAt, equalTo(5L));
+        task.punctuate(processor, 10);
+        assertThat(processor.punctuatedAt, equalTo(10L));
+    }
+
+    @Test
+    public void shouldSetProcessorNodeOnContextBackToNullAfterSuccesfullPunctuate() throws Exception {
+        task.punctuate(processor, 5);
+        assertThat(((ProcessorContextImpl) task.processorContext()).currentNode(), nullValue());
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void shouldThrowIllegalStateExceptionOnScheduleIfCurrentNodeIsNull() throws Exception {
+        task.schedule(1);
+    }
+
+    @Test
+    public void shouldNotThrowIExceptionOnScheduleIfCurrentNodeIsNotNull() throws Exception {
+        ((ProcessorContextImpl) task.processorContext()).setCurrentNode(processor);
+        task.schedule(1);
+    }
+
+    @SuppressWarnings("unchecked")
+    @Test
+    public void shouldThrowExceptionIfAnyExceptionsRaisedDuringCloseTopology() throws Exception {
+        final MockSourceNode processorNode = new MockSourceNode(topic1, intDeserializer, intDeserializer) {
+            @Override
+            public void close() {
+                throw new RuntimeException("KABOOM!");
+            }
+        };
+        final List<ProcessorNode> processorNodes = Collections.<ProcessorNode>singletonList(processorNode);
+        final Map<String, SourceNode> sourceNodes
+                = Collections.<String, SourceNode>singletonMap(topic1[0], processorNode);
+        final ProcessorTopology topology = new ProcessorTopology(processorNodes,
+                                                                 sourceNodes,
+                                                                 Collections.<String, SinkNode>emptyMap(),
+                                                                 Collections.<StateStore>emptyList(),
+                                                                 Collections.<String, String>emptyMap(),
+                                                                 Collections.<StateStore>emptyList());
+
+
+        task = new StreamTask(taskId00, applicationId, partitions,
+                              topology, consumer, restoreStateConsumer, config, streamsMetrics, stateDirectory, testCache, time, recordCollector);
+
+        try {
+            task.closeTopology();
+            fail("should have thrown runtime exception");
+        } catch (RuntimeException e) {
+            // ok
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    @Test
+    public void shouldCloseAllProcessorNodesWhenExceptionsRaised() throws Exception {
+        final MockSourceNode processorNode = new MockSourceNode(topic1, intDeserializer, intDeserializer) {
```

Comment:
Unify "create task" code with `shouldThrowExceptionIfAnyExceptionsRaisedDuringCloseTopology` -- it's almost the same and both test cases can use the same topology structure.